man/u_man/man1/rcsmerge.1
'\"macro stdmacro
.TH RCSMERGE 1
.SH NAME
rcsmerge \- merge RCS revisions
.SH SYNOPSIS
\f3rcsmerge\f1 \f3\-r\f2rev1\f1 [ \f3\-r\f2rev2\f1 ] [ \f3\-p\f1 ] file
.SH DESCRIPTION
.I rcsmerge
incorporates the changes between \f2rev1\f1 and \f2rev2\f1 of an
RCS file into the corresponding working file. If \f3\-p\f1 is given, the result
is printed on the std. output, otherwise the result overwrites the
working file.
.PP
A file name ending in ',v' is an RCS file name, otherwise a
working file name. \f2merge\f1 derives the working file name from the RCS
file name and vice versa, as explained in \f2co\f1(1). A pair consisting
of both an RCS and a working file name may also be specified.
.PP
\f2rev1\f1 may not be omitted. If \f2rev2\f1 is omitted, the latest
revision on the trunk is assumed.
Both \f2rev1\f1 and \f2rev2\f1 may be given numerically or symbolically.
.PP
\f2rcsmerge\f1 prints a warning if there are overlaps, and delimits
the overlapping regions as explained in the \f3\-j\f1 version of
\f2co\f1.
The command is useful for incorporating changes into a checked-out revision.
.SH EXAMPLES
Suppose you have released revision 2.8 of f.c. Assume
furthermore that you just completed revision 3.4, when you receive
updates to release 2.8 from someone else.
To combine the updates to 2.8 and your changes between 2.8 and 3.4,
put the updates to 2.8 into file f.c and execute
.sp
        rcsmerge  \-p  \-r2.8  \-r3.4  f.c  >f.merged.c
.sp
Then examine f.merged.c.
Alternatively, if you want to save the updates to 2.8 in the RCS file,
check them in as revision 2.8.1.1 and execute \f2co \-j\f1:
.sp
        ci  \-r2.8.1.1  f.c
        co  \-r3.4  \-j2.8:2.8.1.1  f.c
.sp
As another example, the following command undoes the changes
between revision 2.4 and 2.8 in your currently checked out revision
in f.c.
.sp
        rcsmerge  \-r2.8  \-r2.4  f.c
.sp
Note the order of the arguments, and that f.c will be
overwritten.
.SH IDENTIFICATION
Author: Walter F. Tichy,
.br
Purdue University, West Lafayette, IN, 47907.
.br
Copyright \(co 1982 by Walter F. Tichy.
.SH SEE ALSO
ci(1), co(1), merge(1), ident(1), rcs(1), rcsdiff(1), rlog(1), rcsfile(4).
.sp 0
Walter F. Tichy, ``Design, Implementation, and Evaluation of a Revision Control
System,'' in \f2Proceedings of the 6th International Conference on Software
Engineering\f1, IEEE, Tokyo, Sept. 1982.
.SH BUGS
\f2rcsmerge\f1 does not work for
files that contain lines with a single `.'.


usr.bin/rcs/src/rcs/rcsmerge.c
/*
 *                       rcsmerge operation
 */
 static char rcsid[]=
 "$Header: /d2/3.7/src/usr.bin/rcs/src/rcs/RCS/rcsmerge.c,v 1.1 89/03/27 18:21:51 root Exp $ Purdue CS";
/*****************************************************************************
 *                       join 2 revisions with respect to a third
 *****************************************************************************
 *
 * Copyright (C) 1982 by Walter F. Tichy
 *                       Purdue University
 *                       Computer Science Department
 *                       West Lafayette, IN 47907
 *
 * All rights reserved. No part of this software may be sold or distributed
 * in any form or by any means without the prior written permission of the
 * author.
 * Report problems and direct all inquiries to Tichy@purdue (ARPA net).
 */


/* $Log:	rcsmerge.c,v $
 * Revision 1.1  89/03/27  18:21:51  root
 * Initial check-in for 3.7
 * 
 * Revision 1.1  85/10/30  16:09:52  brendan
 * Initial revision
 * 
 * Revision 3.4  83/09/09  13:44:15  ellis
 * ported to SGI
 * 
 * Revision 3.3  82/12/24  15:29:00  wft
 * Added call to catchsig().
 * 
 * Revision 3.2  82/12/10  21:32:02  wft
 * Replaced getdelta() with gettree(); improved error messages.
 *
 * Revision 3.1  82/11/28  19:27:44  wft
 * Initial revision.
 *
 */
#include "rcsbase.h"
static char rcsbaseid[] = RCSBASE;

extern int  cleanup();              /* cleanup after signals                */
extern char * mktempfile();         /*temporary file name generator         */
extern struct hshentry * genrevs(); /*generate delta numbers                */
extern int  nerror;                 /*counter for errors                    */

char *RCSfilename;
char *workfilename;
char * temp1file, * temp2file;

main (argc, argv)
int argc; char **argv;
{
        char * cmdusage;
        char command[NCPPN+revlength+40];
        int  revnums; /* counter for revision numbers given */
        int tostdout;
        char * rev1, * rev2; /*revision numbers*/
        char numericrev[revlength];   /* holds expanded revision number     */
        struct hshentry * gendeltas[hshsize];/*stores deltas to be generated*/
        struct hshentry * target;

        catchints();
        cmdid = "rcsmerge";
        cmdusage = "command format:\n    rcsmerge -p -rrev1 -rrev2 file\n    rcsmerge -p -rrev1 file";
        revnums=0;tostdout=false;

        while (--argc,++argv, argc>=1 && ((*argv)[0] == '-')) {
                switch ((*argv)[1]) {
                case 'p':
                        tostdout=true;
                        /* falls into -r */
                case 'r':
                        if ((*argv)[2]!='\0') {
                            if (revnums==0) {
                                    rev1= *argv+2; revnums=1;
                            } elif (revnums==1) {
                                    rev2= *argv+2; revnums=2;
                            } else {
                                    faterror("too many revision numbers");
                            }
                        } /* do nothing for empty -r or -p */
                        break;

                default:
                        faterror("unknown option: %s\n%s", *argv,cmdusage);
                };
        } /* end of option processing */

        if (argc<1) faterror("No input file\n%s",cmdusage);
        if (revnums<1) faterror("no base revision number given");

        /* now handle all filenames */

        if (pairfilenames(argc,argv,true,false)==1) {

                if (argc>2 || (argc==2&&argv[1]!=nil))
                        warn("too many arguments");
                diagnose("RCS file: %s",RCSfilename);
                if (!(access(workfilename,4)==0))
                        faterror("Can't open %s",workfilename);

                if (!trysema(RCSfilename,false)) goto end; /* give up */

                gettree();  /* reads in the delta tree */

                if (Head==nil) faterror("no revisions present");


                if (!expandsym(rev1,numericrev)) goto end;
                if (!(target=genrevs(numericrev,nil,nil,nil,gendeltas))) goto end;
                rev1=target->num;
                if (revnums==1) rev2=Head->num; /* default for rev2 */
                if (!expandsym(rev2,numericrev)) goto end;
                if (!(target=genrevs(numericrev,nil,nil,nil,gendeltas))) goto end;
                rev2=target->num;

                temp1file=mktempfile("/tmp/",TMPFILE1);
                temp2file=mktempfile("/tmp/",TMPFILE2);

                diagnose("retrieving revision %s",rev1);
                sprintf(command,"%s/co -q -p%s %s > %s\n",
                        TARGETDIR,rev1,RCSfilename,temp1file);
                if (system(command)){
                        faterror("co failed");
                }
                diagnose("retrieving revision %s",rev2);
                sprintf(command,"%s/co -q -p%s %s > %s\n",
                         TARGETDIR,rev2,RCSfilename,temp2file);
                if (system(command)){
                        faterror("co failed");
                }
                diagnose("Merging differences between %s and %s into %s%s",
                         rev1, rev2, workfilename,
                         tostdout?"; result to stdout":"");

                sprintf(command,"%s %s%s %s %s %s %s\n",MERGE,tostdout?"-p ":"",
                        workfilename,temp1file,temp2file,workfilename,rev2);
                if (system(command)) {
                        faterror("merge failed");
                }
        }

end:
        cleanup();
        exit(nerror!=0);

}

